### DISCUSSION TITLE
Some Lua pattern in Teal.

### DISCUSSION THREAD

**Discussion by nagolove** [teal-language/tl#350] [Q&A] [answered]
In my project I have next pattern.
```lua
local commands = {}

commands["do3"] = function()
    print("do3")
end

function commands.do1() print("do1") end
function commands.do2() print("do2") end

function process(cmd)
    if cmd then
        local command = commands[cmd]
        if command then
            command()
        end
    end
end
```
How I can rewrite such code?
First attempt
```lua
local commands: {string: function} = {}

commands["do3"] = function()
    print("do3")
end

function commands.do1() print("do1") end
function commands.do2() print("do2") end

function process(cmd: string)
    if cmd then
        local command = commands[cmd]
        if command then
            command()
        end
    end
end
```
It has `not a module: {string : function(<any type>):<any type>}` error in `function commands.do1() print("do1") end` and next function. But it has no error in `commands["do3"] = function()`. Why?
Next attempt. More type declarations.
```lua
local type Commands = record
    do1: function()
    do2: function()
    do3: function()
end

local type Cmds = enum
    "do1"
    "do2"
    "do3"
end

local commands: Commands = {}

commands["do3"] = function()
    print("do3")
end

function commands.do1() print("do1") end
function commands.do2() print("do2") end

function process(cmd: Cmds)
    if cmd then
        local command: function() = commands[cmd]
        if command then
            command()
        end
    end
end
```
It compiles but has more duplication. Also I am embarrassed with types naming for record and enum. What is right approach for such patterns?

**Comment by lenscas** [comment, marked as answer]
every function that can be executed using that index need to have the same type, so what you can do is something like
```lua
local type CommandNames = enum
    "do1"
    "do2"
end

local commands : {CommandNames : function()} = { 
 do1 = function() print("do1")end,
 do2 = function() print("do2")end
}

function process(cmd: CommandNames)
    if cmd then
        local command: function() = commands[cmd]
        if command then
            command()
        end
    end
end
```

Basically, stop using a record and start using a map of commandName : yourFunctionType

**Comment by hishamhm** [comment]
@lenscas's response is correct, this pattern calls for a map, not a record.

Which I noticed was the first thing you tried to do, but stumbled into this:

> It has not a module: {string : function(<any type>):<any type>} error in function commands.do1() print("do1") end and next function. But it has no error in commands["do3"] = function(). Why?

That's a good question. Lua supports different syntaxes for some equivalent operations on tables, because the table type takes on many duties (maps, records, arrays, etc.) — in Teal, I have reserved the `function r.f()` syntax for records and the assignment syntax `r.f = function()` for maps.

The reason is inference: a typical pattern is to do

```
local mymodule = {}

function mymodule.myfunction()
   -- end
end
```

so Teal infers that `mymodule` is a record and `function mymodule.myfunction()` is _declaring_ a field to that record.

Since you have declared in your first attempt that you wanted `commands` to be a map _of strings to functions_, I guess Teal could make an exception and accept that syntax (but noting that the function signature would have to match that of the map value type exactly).

But a slight variation of your first attempt does, in fact, work:

```lua
local commands: {string: function} = {}

commands["do3"] = function()
    print("do3")
end
commands["do1"] = function()
    print("do1")
end
commands["do2"] = function()
    print("do2")
end

function process(cmd: string)
    if cmd then
        local command = commands[cmd]
        if command then
            command()
        end
    end
end
```

**Reply by lenscas** [reply]
might be useful to note that using an `enum` gives you better type safety, as it will be impossible to give `process` a cmd key that isn't valid :)
